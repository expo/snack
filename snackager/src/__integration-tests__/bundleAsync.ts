import fs from 'fs';
import glob from 'glob';
import mapValues from 'lodash/mapValues';
import os from 'os';
import path from 'path';
import util from 'util';

import { getCoreExternals } from '../bundler/externals';
import fetchAndExtract from '../utils/fetchAndExtract';
import fetchMetadata from '../utils/fetchMetadata';
import findVersion from '../utils/findVersion';
import getBundleInfo, { BundleInfo } from '../utils/getBundleInfo';
import installPackage from '../utils/installPackage';
import packageBundle from '../utils/packageBundle';
import parseRequest from '../utils/parseRequest';
import resolveDependencies from '../utils/resolveDependencies';

// TODO: replace rimraf with fs.rm once node 14.40 lands
const rimraf = util.promisify(require('rimraf'));

const DEFAULT_PLATFORMS = ['ios', 'android', 'web'];

type BundledPackage = {
  name: string;
  version: string;
  // TODO: fix possible null dependency and replace with `Package['peerDependencies']`
  peerDependencies: { [dependency: string]: string | null };
  files: { [platform: string]: { [file: string]: BundleInfo } };
};

// Keep track of all tested packages, to clean the lockfiles if they aren't tested anymore
const testedPackages: string[] = [];

export default async function bundleAsync(
  packageSpec: string,
  bundlePlatforms: string[] = DEFAULT_PLATFORMS,
  includeCode?: boolean
): Promise<BundledPackage> {
  const { qualified, id, tag, scope, deep, platforms } = parseRequest(
    `/${packageSpec}?platforms=${bundlePlatforms.join(',')}`
  );
  const workdir = path.join(
    fs.realpathSync(os.tmpdir()),
    'snackager',
    '__integration-tests__',
    qualified
  );
  try {
    await rimraf(workdir);
  } catch (e) {}
  const packagedir = path.join(workdir, 'package');

  const meta = await fetchMetadata(qualified, { scope, id });
  const { version, isLatest } = findVersion(qualified, meta, tag);
  const { pkg, dependencies } = resolveDependencies(meta, version, isLatest, deep);

  await fetchAndExtract(pkg, version, workdir);

  // keep track of the used lockfiles, and try to restore the lockfile
  // this represents the package + nested dependency "state" that was used in the bundle snapshot
  testedPackages.push(packageSpec);
  const cacheHit = await restoreLockfile(packageSpec, packagedir);
  if (cacheHit) {
    await installPackage(packagedir, ['--frozen-lockfile']);
  } else {
    await installPackage(packagedir);
    await saveLockfile(packageSpec, packagedir);
  }

  const files = await packageBundle({
    pkg,
    cwd: packagedir,
    platforms,
    base: 'nonsense',
    deep,
    externalDependencies: dependencies,
  });

  return {
    name: qualified,
    version,
    files: mapValues(files, (file, platform) =>
      mapValues(file, (buffer, filename) => {
        const metadata = getBundleInfo(filename, buffer, includeCode);
        (metadata.externals ?? []).forEach((external) => {
          if (!getCoreExternals().includes(external) && !dependencies[external]) {
            console.warn(
              `Bundle "${qualified}@${version}/${platform}-${filename}" contains external "${external}" which is not listed as a peer dependency in package.json`
            );
          }
        });
        return metadata;
      })
    ),
    peerDependencies: dependencies,
  };
}

// Location of the lockfiles to store them in, they should be considered similar to snapshots
const lockfileStorage = path.join(__dirname, '__snapshots__', 'lockfiles');

async function restoreLockfile(packageSpec: string, directory: string): Promise<boolean> {
  const packagePath = path.join(directory, 'yarn.lock');
  const storagePath = path.join(lockfileStorage, `${packageSpec}.lock`);

  try {
    await fs.promises.copyFile(storagePath, packagePath);
    // Lockfile is restored successfully (cache hit)
    return true;
  } catch {
    // Lockfile couldn't be restored (no cache hit)
    return false;
  }
}

async function saveLockfile(packageSpec: string, directory: string): Promise<void> {
  const packagePath = path.join(directory, 'yarn.lock');
  const storagePath = path.join(lockfileStorage, `${packageSpec}.lock`);

  await fs.promises.mkdir(path.dirname(storagePath), { recursive: true });
  // Always keep the lockfile up to date after an install
  await fs.promises.copyFile(packagePath, storagePath);

  console.warn(`Added lockfile for ${packageSpec}, commit this to the repository.`);
}

export async function cleanUnusedLockfiles(): Promise<void> {
  const existingPackages = await new Promise<string[]>((resolve, reject) => {
    glob('**/*.lock', { cwd: lockfileStorage }, (error, files) => {
      if (error) {
        reject(error);
      } else {
        resolve(files.map((filename) => filename.replace('.lock', '')));
      }
    });
  });

  const unusedPackages = existingPackages.filter(
    (packageSpec) => !testedPackages.includes(packageSpec)
  );

  // remove unused package lockfiles from the repo
  for (const packageSpec of unusedPackages) {
    await fs.promises.unlink(path.join(lockfileStorage, `${packageSpec}.lock`));
    console.warn(`Removed obsolete lockfile for ${packageSpec}, commit this to the repository.`);
  }
}
