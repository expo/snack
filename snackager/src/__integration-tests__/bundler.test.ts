import bundleAsync, { cleanUnusedLockfiles } from './bundleAsync';

jest.setTimeout(1000 * 60 * 2);

// This integration test needs to validate if the packages are being bundled properly.
// To validate `bundle.js` from `<package>@<version`, we need to use the bundle size.
// It's an indicator of the bundled content, e.g. when it's 0 something is wrong

// Unfortunately, when a nested dependency is updated the bundled code changes in size.
// To prevent raising false alerts, we need to manage lockfiles for each tested `<package>@<version>`.
//   - Lockfiles are generated by Yarn and copied to `__snapshots__/lockfiles` by `bundleAsync` (after installing)
//   - When we have an existing lockfile, `bundleAsync` restores it just before installing
//   - `bundleAsync` keeps track of all used lockfiles, to auto-clean the repo when tests changes.
afterAll(() => cleanUnusedLockfiles());

describe('bundler', () => {
  it('creates bundle for a single platform', async () => {
    const bundle = await bundleAsync('firestorter@2.0.1', ['ios']);
    expect(bundle).toMatchSnapshot();
  });

  it('fails when no package name is specified', async () => {
    await expect(bundleAsync('', ['ios'])).rejects.toEqual(new Error(`Failed to parse request`));
  });

  it('externalizes dependencies that are marked as external', async () => {
    // `expo-google-app-auth` declares `expo-app-auth` as a direct dependency.
    // The `expo-app-auth` dependency should however still be externalized and
    // not linked into the bundle.
    const bundle = await bundleAsync('expo-google-app-auth@8.1.3');
    expect(bundle).toMatchSnapshot();
    expect(bundle.files.android['bundle.js'].externals).toEqual(
      expect.arrayContaining(['expo-app-auth'])
    );
  });

  it('resolves external source-code references', async () => {
    // `expo-linear-gradient` imports source-code from react-native-web directly.
    // e.g. "import normalizeColor from 'react-native-web/src/modules/normalizeColor';"
    // The bundler should install `react-native-web` and resolve the import correctly.
    // All regular `react-native-web` imports should be externalized.
    const bundle = await bundleAsync('expo-linear-gradient@8.2.1');
    expect(bundle).toMatchSnapshot();
  });

  it('resolves commonly missing peer-deps (prop-types)', async () => {
    // For example 'react-native-responsive-grid@0.32.4' has a dependency
    // on prop-types but no peerDep listed in package.json.
    const bundle = await bundleAsync('react-native-responsive-grid@0.32.4');
    expect(bundle).toMatchSnapshot();
    expect(bundle.files.android['bundle.js'].externals).not.toEqual(
      expect.arrayContaining(['prop-types'])
    );
  });

  it.skip('creates bundle for @react-navigation/native', async () => {
    const bundle = await bundleAsync('@react-navigation/native@5.7.3');
    expect(bundle).toMatchSnapshot();
    // @react-navigation/core should be included in the bundle and not an external
    expect(bundle.files.android['bundle.js'].externals).not.toEqual(
      expect.arrayContaining(['@react-navigation/core'])
    );
  });

  it('creates bundle for @react-navigation/stack', async () => {
    const bundle = await bundleAsync('@react-navigation/stack@5.9.0');
    expect(bundle).toMatchSnapshot();
    // @react-navigation/native should be included as an external
    expect(bundle.files.android['bundle.js'].externals).toEqual(
      expect.arrayContaining(['@react-navigation/native'])
    );
  });

  it('externalizes references to react-native-gesture-handler/DrawerLayout', async () => {
    const bundle = await bundleAsync('react-navigation@3.13.0', ['ios']);
    expect(bundle).toMatchSnapshot();
    // react-native-gesture-handler/DrawerLayout should be included as an external
    expect(bundle.files.ios['bundle.js'].externals).toEqual(
      expect.arrayContaining(['react-native-gesture-handler/DrawerLayout'])
    );
  });

  it('creates bundle for subpath', async () => {
    const bundle = await bundleAsync('react-native-gesture-handler/DrawerLayout@1.6.0');
    expect(bundle).toMatchSnapshot();
  });

  it('disallows bundling of core modules', async () => {
    await expect(bundleAsync('expo', ['ios'])).rejects.toEqual(
      new Error(`Bundling core module 'expo' is prohibited`)
    );
    await expect(bundleAsync('react-native', ['ios'])).rejects.toEqual(
      new Error(`Bundling core module 'react-native' is prohibited`)
    );
    await expect(bundleAsync('react-native-web', ['ios'])).rejects.toEqual(
      new Error(`Bundling core module 'react-native-web' is prohibited`)
    );
    await expect(bundleAsync('react-native-windows', ['ios'])).rejects.toEqual(
      new Error(`Bundling core module 'react-native-windows' is prohibited`)
    );
    await expect(
      bundleAsync('react-native/Libraries/Image/AssetRegistry', ['ios'])
    ).rejects.toEqual(
      new Error(`Bundling core module 'react-native/Libraries/Image/AssetRegistry' is prohibited`)
    );
  });

  it('creates bundle for subpath of core package', async () => {
    const bundle2 = await bundleAsync('react-native-web/src/modules/normalizeColor@0.14.4');
    expect(bundle2).toMatchSnapshot();
  });

  it('created bundle for react-native-gesture-handler', async () => {
    const bundle = await bundleAsync('react-native-gesture-handler@1.6.0');
    expect(bundle).toMatchSnapshot();
  });

  it('creates bundle for react-native-webview', async () => {
    const bundle = await bundleAsync('react-native-webview@10.9.1');
    expect(bundle).toMatchSnapshot();
    // react-native-webview contains a direct reference to
    // 'react-native/Libraries/BatchedBridge/BatchedBridge` and
    // should bundle correctly
  });

  it('creates bundle for react-native-screens/native-stack', async () => {
    const bundle = await bundleAsync('react-native-screens/native-stack@2.11.0');
    expect(bundle).toMatchSnapshot();
    // react-native-webview contains a direct reference to
    // 'react-native/Libraries/ReactNative/AppContainer` and
    // should bundle correctly
  });

  it('creates native bundles when web entry point is not found', async () => {
    const bundle = await bundleAsync('react-native-reanimated@2.0.0-alpha.6');
    // entry point cannot be found for the web platform for this package.
    // this should not cause the bundler to fail but continue with the native
    // platforms instead.
    expect(bundle.files['ios']['bundle.js'].size).toBeGreaterThanOrEqual(10000);
    expect(bundle.files['android']['bundle.js'].size).toBeGreaterThanOrEqual(10000);
    expect(bundle.files['web']['bundle.js'].size).toBe(0);
  });

  it('compiles reanimated2 worklets', async () => {
    const bundle = await bundleAsync('react-native-reanimated@2.0.0-rc.3', undefined, true);
    // verify that the number of expected worklet instances appear in the
    // generated bundle.
    expect(bundle.files['ios']['bundle.js'].code!.match(/worklet/g)!.length).toBe(150);
    expect(bundle.files['android']['bundle.js'].code!.match(/worklet/g)!.length).toBe(150);
  });

  it('compiles packages that use reanimated2', async () => {
    const bundle = await bundleAsync('moti@0.10.0', undefined, true);
    ['ios', 'android', 'web'].forEach((platform) => {
      expect(bundle.files[platform]['bundle.js'].code!.match(/worklet/g)!.length).toBe(11);
      expect(bundle.files[platform]['bundle.js'].size).toBeLessThanOrEqual(200000);
      expect(bundle.files[platform]['bundle.js'].externals).toEqual(
        expect.arrayContaining(['react-native-reanimated'])
      );
    });
  });

  it('filter aliased react-native dependencies', async () => {
    const bundle = await bundleAsync('@react-native-community/datetimepicker@3.0.3');
    expect(bundle).toMatchSnapshot();
    // datetimepicker contains a peer-dependency on `react-native-windows`
    // which causes the snack-sdk to try and bundle that dependency as it is not
    // listed as a preloaded module. 'react-native-windows' is considered a special
    // alias for 'react-native` and it therefore removed as a peer dependency
    expect(bundle.peerDependencies['react-native-windows']).toBeUndefined();
  });

  it(`filter internal react-native dependency TextInputState`, async () => {
    const bundle = await bundleAsync('@react-native-community/viewpager@4.2.0');
    // this version detects TextInputState as missing package on `react-native@^0.57.0`
    // validate that we can install and bundle it properly
    expect(bundle).toMatchSnapshot();
    expect(bundle.peerDependencies['TextInputState']).toBeUndefined();
  });

  it('externalizes package-externals of dependencies', async () => {
    const bundle = await bundleAsync('@draftbit/ui@40.34.4');
    // @drafbit/ui depends on @drafbit/core, which in turn depends
    // on `react-native-safe-area-context`. `react-native-safe-area-context` is
    // a package dependency which should be excluded from the bundle, even
    // if the import happens in a dependency of the requested package
    expect(bundle).toMatchSnapshot();
    ['ios', 'android', 'web'].forEach((platform) => {
      expect(bundle.files[platform]['bundle.js'].externals).toEqual(
        expect.arrayContaining(['react-native-safe-area-context'])
      );
    });
  });

  it('creates bundle for graphql@14.0.0', async () => {
    // This library was causing issues in aws-amplify because of mjs files.
    // It's imported by @aws-amplify/api -> @aws-amplify/api-graphql -> graphql,
    // because aws-amplify is HUGE, we are only testing `graphql@14.0.0` here.
    // see: https://github.com/graphql/graphql-js/issues/1272
    const bundle = await bundleAsync('graphql@14.0.0');
    expect(bundle).toMatchSnapshot();
    // Also validate that we don't have any relative extenals
    ['android', 'ios', 'web'].forEach((platform) => {
      expect(bundle.files[platform]['bundle.js'].externals).not.toEqual(
        expect.arrayContaining([expect.stringContaining('./')])
      );
    });
  });

  it('creates bundle for graphql@15.5.1', async () => {
    // This version of graphql contains a fix for the issue in graphql@14.
    // In this version, mjs files have the `.mjs` extension when imported.
    const bundle = await bundleAsync('graphql@15.5.1');
    expect(bundle).toMatchSnapshot();
    // Also validate that we don't have any relative externals
    ['android', 'ios', 'web'].forEach((platform) => {
      expect(bundle.files[platform]['bundle.js'].externals).not.toEqual(
        expect.arrayContaining([expect.stringContaining('./')])
      );
    });
  });

  it('bundles packages with json imports', async () => {
    const bundle = await bundleAsync('react-native-svg@12.1.1');
    expect(bundle).toMatchSnapshot();
  });
});
