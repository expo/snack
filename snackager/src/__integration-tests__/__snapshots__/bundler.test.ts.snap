// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundler created bundle for react-native-gesture-handler 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
        ],
        "size": 69283,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "react-native",
          "react",
        ],
        "size": 66895,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
        ],
        "size": 138143,
      },
    },
  },
  "name": "react-native-gesture-handler",
  "peerDependencies": Object {},
  "version": "1.6.0",
}
`;

exports[`bundler creates bundle for @react-navigation/native 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
        ],
        "size": 102430,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
        ],
        "size": 102426,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
        ],
        "size": 107732,
      },
    },
  },
  "name": "@react-navigation/native",
  "peerDependencies": Object {
    "react": "*",
    "react-native": "*",
  },
  "version": "5.7.3",
}
`;

exports[`bundler creates bundle for @react-navigation/stack 1`] = `
Object {
  "files": Object {
    "android": Object {
      "assets/src_views_assets_backicon.png": Object {
        "size": 100,
      },
      "assets/src_views_assets_backicon@1.5x.png": Object {
        "size": 134,
      },
      "assets/src_views_assets_backicon@2x.png": Object {
        "size": 134,
      },
      "assets/src_views_assets_backicon@3x.png": Object {
        "size": 167,
      },
      "assets/src_views_assets_backicon@4x.png": Object {
        "size": 207,
      },
      "assets/src_views_assets_backiconmask.png": Object {
        "size": 913,
      },
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
          "@react-navigation/native",
          "react-native-safe-area-context",
          "AssetRegistry",
          "react-native-gesture-handler",
          "react-native-screens",
          "@react-native-community/masked-view",
        ],
        "size": 100808,
      },
    },
    "ios": Object {
      "assets/src_views_assets_backicon.png": Object {
        "size": 290,
      },
      "assets/src_views_assets_backicon@1.5x.png": Object {
        "size": 373,
      },
      "assets/src_views_assets_backicon@2x.png": Object {
        "size": 405,
      },
      "assets/src_views_assets_backicon@3x.png": Object {
        "size": 761,
      },
      "assets/src_views_assets_backicon@4x.png": Object {
        "size": 812,
      },
      "assets/src_views_assets_backiconmask.png": Object {
        "size": 913,
      },
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
          "@react-navigation/native",
          "react-native-safe-area-context",
          "react-native-gesture-handler",
          "AssetRegistry",
          "react-native-screens",
          "@react-native-community/masked-view",
        ],
        "size": 101787,
      },
    },
    "web": Object {
      "assets/lib_module_views_assets_backicon.png": Object {
        "size": 207,
      },
      "assets/lib_module_views_assets_backiconmask.png": Object {
        "size": 913,
      },
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
          "@react-navigation/native",
          "react-native-safe-area-context",
          "AssetRegistry",
          "react-native-screens",
        ],
        "size": 105768,
      },
    },
  },
  "name": "@react-navigation/stack",
  "peerDependencies": Object {
    "@react-native-community/masked-view": "^0.1.0",
    "@react-navigation/native": "5.0.5",
    "react": "*",
    "react-native": "*",
    "react-native-gesture-handler": "^1.0.0",
    "react-native-safe-area-context": "^0.6.0",
    "react-native-screens": "^2.0.0",
  },
  "version": "5.9.0",
}
`;

exports[`bundler creates bundle for a single platform 1`] = `
Object {
  "files": Object {
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "mobx",
        ],
        "size": 44972,
      },
    },
  },
  "name": "firestorter",
  "peerDependencies": Object {
    "mobx": "^4.0.0",
  },
  "version": "2.0.1",
}
`;

exports[`bundler creates bundle for react-native-screens/native-stack 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [
          "@react-navigation/native",
          "react",
          "react-native-screens",
          "react-native",
          "react-native/Libraries/ReactNative/AppContainer",
        ],
        "size": 9284,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "@react-navigation/native",
          "react",
          "react-native-screens",
          "react-native",
          "react-native/Libraries/ReactNative/AppContainer",
        ],
        "size": 9280,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [
          "@react-navigation/native",
          "react",
          "react-native-screens",
          "react-native",
          "react-native/Libraries/ReactNative/AppContainer",
        ],
        "size": 9115,
      },
    },
  },
  "name": "react-native-screens",
  "peerDependencies": Object {
    "react": "*",
    "react-native": "*",
  },
  "version": "2.11.0",
}
`;

exports[`bundler creates bundle for react-native-webview 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [
          "react-native",
          "react",
          "react-native/Libraries/BatchedBridge/BatchedBridge",
        ],
        "size": 12236,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "react-native",
          "react",
        ],
        "size": 11901,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
        ],
        "size": 1684,
      },
    },
  },
  "name": "react-native-webview",
  "peerDependencies": Object {
    "react-native": "0.64.0",
  },
  "version": "10.9.1",
}
`;

exports[`bundler creates bundle for subpath 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [
          "react-native",
          "react",
        ],
        "size": 46834,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "react-native",
          "react",
        ],
        "size": 46830,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
        ],
        "size": 118074,
      },
    },
  },
  "name": "react-native-gesture-handler",
  "peerDependencies": Object {},
  "version": "1.6.0",
}
`;

exports[`bundler creates bundle for subpath of core package 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [],
        "size": 6865,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [],
        "size": 6861,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [],
        "size": 6861,
      },
    },
  },
  "name": "react-native-web",
  "peerDependencies": Object {
    "react": "^16.5.1",
    "react-dom": "^16.5.1",
  },
  "version": "0.14.4",
}
`;

exports[`bundler externalizes dependencies that are marked as external 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [
          "@unimodules/core",
          "expo-app-auth",
          "expo-constants",
          "react-native",
        ],
        "size": 14413,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "@unimodules/core",
          "expo-app-auth",
          "expo-constants",
          "react-native",
        ],
        "size": 14409,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [
          "@unimodules/core",
          "expo-app-auth",
          "expo-constants",
          "react-native",
        ],
        "size": 14409,
      },
    },
  },
  "name": "expo-google-app-auth",
  "peerDependencies": Object {
    "@unimodules/core": "*",
    "expo-app-auth": "~9.2.0",
    "expo-constants": "*",
    "react": "*",
    "react-native": "*",
  },
  "version": "8.1.3",
}
`;

exports[`bundler externalizes references to react-native-gesture-handler/DrawerLayout 1`] = `
Object {
  "files": Object {
    "ios": Object {
      "assets/node_modules_reactnavigationstack_lib_module_views_assets_backicon.png": Object {
        "size": 290,
      },
      "assets/node_modules_reactnavigationstack_lib_module_views_assets_backicon@1.5x.png": Object {
        "size": 373,
      },
      "assets/node_modules_reactnavigationstack_lib_module_views_assets_backicon@2x.png": Object {
        "size": 405,
      },
      "assets/node_modules_reactnavigationstack_lib_module_views_assets_backicon@3x.png": Object {
        "size": 761,
      },
      "assets/node_modules_reactnavigationstack_lib_module_views_assets_backicon@4x.png": Object {
        "size": 812,
      },
      "assets/node_modules_reactnavigationstack_lib_module_views_assets_backiconmask.png": Object {
        "size": 1619,
      },
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
          "@react-navigation/native",
          "react-native-screens",
          "react-native-gesture-handler",
          "AssetRegistry",
          "react-native-gesture-handler/DrawerLayout",
        ],
        "size": 221207,
      },
    },
  },
  "name": "react-navigation",
  "peerDependencies": Object {
    "@react-navigation/native": "~3.6.2",
    "react": "*",
    "react-native": "*",
  },
  "version": "3.13.0",
}
`;

exports[`bundler filter aliased react-native dependencies 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [
          "react-native",
        ],
        "size": 13228,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "react-native",
          "react",
        ],
        "size": 9063,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
        ],
        "size": 1565,
      },
    },
  },
  "name": "@react-native-community/datetimepicker",
  "peerDependencies": Object {
    "react": "^16.8.3",
    "react-native": "^0.59.0",
  },
  "version": "3.0.3",
}
`;

exports[`bundler filter internal react-native dependency TextInputState 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
          "react-native/Libraries/Utilities/dismissKeyboard",
        ],
        "size": 8187,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
          "react-native/Libraries/Utilities/dismissKeyboard",
        ],
        "size": 8183,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
          "react-native/Libraries/Utilities/dismissKeyboard",
        ],
        "size": 8183,
      },
    },
  },
  "name": "@react-native-community/viewpager",
  "peerDependencies": Object {
    "react": "16.0.0",
    "react-native": "^0.57.0",
  },
  "version": "4.2.0",
}
`;

exports[`bundler resolves commonly missing peer-deps (prop-types) 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [
          "react-native",
          "react",
        ],
        "size": 19030,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "react-native",
          "react",
        ],
        "size": 19026,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [
          "react-native",
          "react",
        ],
        "size": 19026,
      },
    },
  },
  "name": "react-native-responsive-grid",
  "peerDependencies": Object {},
  "version": "0.32.4",
}
`;

exports[`bundler resolves external source-code references 1`] = `
Object {
  "files": Object {
    "android": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
          "@unimodules/core",
        ],
        "size": 8091,
      },
    },
    "ios": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
          "@unimodules/core",
        ],
        "size": 7471,
      },
    },
    "web": Object {
      "bundle.js": Object {
        "externals": Array [
          "react",
          "react-native",
        ],
        "size": 14946,
      },
    },
  },
  "name": "expo-linear-gradient",
  "peerDependencies": Object {
    "react": "*",
    "react-native": "*",
  },
  "version": "8.2.1",
}
`;
